scipy/signal/_peak_finding_utils.pyx

```diff
@@ -74,3 +74,98 @@ def _argmaxima1d(np.float64_t[:] x not None):
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+def _peak_prominences(np.float64_t[::1] x not None,
+                      np.intp_t[::1] peaks not None,
+                      np.intp_t wlen):
+    """
+    Calculate the prominence of each peak in a signal.
+
+    Parameters
+    ----------
+    x : ndarray
+        A signal with peaks.
+    peaks : ndarray
+        Indices of peaks in `x`.
+    wlen : np.intp
+        A window length in samples (see `peak_prominences`) which is rounded up
+        to the nearest odd integer. If smaller than 2 the entire signal `x` is
+        used.
+
+    Returns
+    -------
+    prominences : ndarray
+        The calculated prominences for each peak in `peaks`.
+    left_bases, right_bases : ndarray
+        The peaks' bases as indices in `x` to the left and right of each peak.
+
+    Raises
+    ------
+    ValueError
+        If an index in `peaks` doesn't point to a local maximum in `x`.
+
+    Notes
+    -----
+    This is the inner function to `peak_prominences`.
+
+    .. versionadded:: 1.1.0
+    """
+    cdef:
+        np.float64_t[::1] prominences
+        np.intp_t[::1] left_bases, right_bases
+        np.float64_t left_min, right_min
+        np.intp_t peak_nr, peak, i_min, i_max, i
+        bint raise_error
+
+    raise_error = False
+    prominences = np.empty(peaks.shape[0], dtype=np.float64)
+    left_bases = np.empty(peaks.shape[0], dtype=np.intp)
+    right_bases = np.empty(peaks.shape[0], dtype=np.intp)
+
+    with nogil:
+        for peak_nr in range(peaks.shape[0]):
+            peak = peaks[peak_nr]
+            i_min = 0
+            i_max = x.shape[0] - 1
+
+            if 2 <= wlen:
+                # Adjust window around the evaluated peak (within bounds);
+                # if wlen is even the resulting window length is is implicitly
+                # rounded to next odd integer
+                i_min = max(peak - wlen // 2, i_min)
+                i_max = min(peak + wlen // 2, i_max)
+
+            # Find the left base in interval [i_min, peak]
+            i = peak
+            left_min = x[peak]
+            while i_min <= i and x[i] <= x[peak]:
+                if x[i] < left_min:
+                    left_min = x[i]
+                    left_bases[peak_nr] = i
+                i -= 1
+            if not left_min < x[peak]:
+                raise_error = True  # Raise error outside nogil statement
+                break
+
+            # Find the right base in interval [peak, i_max]
+            i = peak
+            right_min = x[peak]
+            while i <= i_max and x[i] <= x[peak]:
+                if x[i] < right_min:
+                    right_min = x[i]
+                    right_bases[peak_nr] = i
+                i += 1
+            if not right_min < x[peak]:
+                raise_error = True  # Raise error outside nogil statement
+                break
+
+            prominences[peak_nr] = x[peak] - max(left_min, right_min)
+
+    if raise_error is True:
```

Comment:
I'm not sure if it affects the generated C code, but it's more pythonic to use `if raise_error:`.

Reply:
I don't have strong feelings on this one but I was under the impression that this is analogous to checking for `None` with the `is` statement. I prefer this style usually because I know I won't be surprised later on by some implicit interpretation of empty lists, etc. 

And indeed it does affect the generated C code:
`if raise_error:` ->  `__pyx_t_13 = (__pyx_v_raise_error != 0); if (__pyx_t_13) {`
`if raise_error is True` -> `__pyx_t_13 = ((__pyx_v_raise_error == 1) != 0); if (__pyx_t_13) {`
I think your way is the better one.

Reply:
In this case you're guaranteed the type of `raise_error` (due to the `cdef` above), so you don't have to worry about implicit weirdness.